beaker_kernel/lib/subkernels/base.py

```diff
@@ -128,3 +130,32 @@ def cleanup(self):
         if self.checkpoints_enabled:
             shutil.rmtree(self.storage_prefix, ignore_errors=True)
             self.checkpoints = []
+
+
+    @tool()
+    async def run_code(self, code: str) -> str:
+        """
+        Execute code in the user's session. After the execution,
+        the state of the kernel will be rolled back to before this tool
+        was used.
+
+        This tool can be help answer questions about the kernel state. For
+        example, a user may ask something about a dictionary `d` and using 
+        run code with the `code` of `d.keys()`.
+
+        This tool can also be used to double check if code will work before 
+        returning it as a final answer.
+
+        Note that this tool does not capture `stdout` AND only returns the
+        results of the last expression evaluated.
+
+        Args:
+            code (str): Code to run directly in Jupyter.
+        Returns:
+            str: Result of the `expr`
+        
+        """
+        checkpoint_index = await self.add_checkpoint()
+        result = await self.evaluate(code)
+        await self.rollback(checkpoint_index)
+        return result
```

Comment:
Unless I'm mistaken, `result` will not be a string.

Reply:
https://github.com/jataware/beaker-kernel/blob/b8bd402d098a46f52167f38fb2987cd4076d7096/beaker_kernel/lib/context.py#L397-L405

https://github.com/jataware/beaker-kernel/blob/b8bd402d098a46f52167f38fb2987cd4076d7096/beaker_kernel/lib/subkernels/python.py#L52-L57

I think evaluate cleans it up for you

Reply:
I don't think so. Have you tested it?

Reply:
Ah, I see. It was working with the LLM so I thought it was fine.

Just tested like this:
```
    @action()
    async def test_eval(self, message):
        """
        """
        a = await self.evaluate("1")
        b = await self.evaluate("'string'")
        c = await self.evaluate("['a', 'b']")
        return [str(type(a)), str(type(b)), str(type(c))]
```

I reread the above I somehow missed line 401. I'll change this to `"return"`. Will resolve in a future proof. Thanks for catching!

Reply:
On top of what you've discovered, keep in mind that the value of "return" will be the actual reified object, not a string. At some point down the line, it will be turned in to a string either via formatting or a direct call to str(return), but keep in mind if you `return result["return"]`, the type returned is `Any`, not `str`.

Reply:
Actually, as I was doing something else, I realized that the agent "wants" more than just the return value. It'll want the stdout and the stderr and probably other things too in addition to the return. Probably best to format it to a string formatted specifically for the agent to use, with headers and seperated sections, etc.